packages/cursorless-engine/src/core/commandVersionUpgrades/generateSpokenForm.ts

```diff
@@ -0,0 +1,478 @@
+            scope,
+          ];
+        }
+        return [ordinalToSpokenForm(modifier.start + 1), scope];
+      }
+
+      const number = numberToSpokenForm(modifier.length);
+
+      if (modifier.start === 0) {
+        return [modifiersExtra.first, number, pluralize(scope)];
+      }
+      if (modifier.start === -modifier.length) {
+        return [modifiersExtra.last, number, pluralize(scope)];
+      }
+
+      throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+    }
+
+    case "range": {
+      if (
+        modifier.anchor.type === "ordinalScope" &&
+        modifier.active.type === "ordinalScope" &&
+        modifier.anchor.length === 1 &&
+        modifier.active.length === 1 &&
+        modifier.anchor.scopeType.type === modifier.active.scopeType.type
+      ) {
+        const anchor =
+          modifier.anchor.start === -1
+            ? modifiersExtra.last
+            : ordinalToSpokenForm(modifier.anchor.start + 1);
+        const active = modifierToSpokenForm(modifier.active);
+        const connective = getRangeConnective(
+          modifier.excludeAnchor,
+          modifier.excludeActive,
+        );
+        return [anchor, connective, active];
+      }
+
+      throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+    }
+
+    default:
+      return [modifiers[modifier.type]];
+  }
+}
+
+function relativeScopeInclusiveToSpokenForm(
+  modifier: RelativeScopeModifier,
+): RecursiveArray<string> {
+  const scope = scopeTypeToSpokenForm(modifier.scopeType);
+
+  if (modifier.length === 1) {
+    const direction =
+      modifier.direction === "forward"
+        ? connectives.forward
+        : connectives.backward;
+
+    // token forward/backward
+    return [scope, direction];
+  }
+
+  const length = numberToSpokenForm(modifier.length);
+  const scopePlural = pluralize(scope);
+
+  // two tokens
+  // This could also have been "two tokens forward"; there is no way to disambiguate.
+  if (modifier.direction === "forward") {
+    return [length, scopePlural];
+  }
+
+  // two tokens backward
+  return [length, scopePlural, connectives.backward];
+}
+
+function relativeScopeExclusiveToSpokenForm(
+  modifier: RelativeScopeModifier,
+): RecursiveArray<string> {
+  const scope = scopeTypeToSpokenForm(modifier.scopeType);
+  const direction =
+    modifier.direction === "forward" ? connectives.next : connectives.previous;
+
+  if (modifier.offset === 1) {
+    const number = numberToSpokenForm(modifier.length);
+
+    if (modifier.length === 1) {
+      // next/previous token
+      return [direction, scope];
+    }
+
+    const scopePlural = pluralize(scope);
+
+    // next/previous two tokens
+    return [direction, number, scopePlural];
+  }
+
+  if (modifier.length === 1) {
+    const ordinal = ordinalToSpokenForm(modifier.offset);
+    // second next/previous token
+    return [ordinal, direction, scope];
+  }
+
+  throw Error(`Modifier '${modifier.type}' is not fully implemented`);
```

Comment:
here as well, I'd think we'd throw `NoSpokenFormError`

Reply:
Again, in this case we could absolutely support `"second next three funks`, so I'd argue that this should be `NoSpokenFormError`.  As mentioned above, if we support it extension side, and it is logically valid, I would argue that `NoSpokenFormError` is appropriate